src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs

```diff
@@ -97,14 +97,15 @@ public CompilationResult Compile([NotNull] IFileInfo fileInfo, [NotNull] string
 
                     if (!result.Success)
                     {
-                        var formatter = new DiagnosticFormatter();
-
-                        var messages = result.Diagnostics
-                                             .Where(IsError)
-                                             .Select(d => GetCompilationMessage(formatter, d))
-                                             .ToList();
-
-                        return CompilationResult.Failed(fileInfo, compilationContent, messages);
+                        var failures = result.Diagnostics.Where(IsError);
+                        var compilationFailure = new RoslynCompilationFailure(failures)
+                        {
+                            CompiledContent = compilationContent,
+                            SourceFileContent = ReadFileContentsSafely(fileInfo.FileInfo),
```

Comment:
We use this to print out the [error stack](https://cloud.githubusercontent.com/assets/174281/5512812/1b287470-8797-11e4-9c68-42a5a79cfb3c.PNG) in Helios \ Diagnostics.